File: scripts/parse-skills.js
const fs = require('fs');
const path = require('path');

const SKILLS_DIR = path.join(__dirname, '..', 'skills');
const OUTPUT_FILE_1 = path.join(__dirname, '..', 'docs', 'architecture', 'codemaps', '003-arch-skills-map.json');
const OUTPUT_FILE_2 = path.join(__dirname, '..', '.agents', 'knowledge', 'kb_skills_mapping.json');

function ensureDirectoryExistence(filePath) {
    const dirname = path.dirname(filePath);
    if (fs.existsSync(dirname)) {
        return true;
    }
    ensureDirectoryExistence(dirname);
    fs.mkdirSync(dirname, { recursive: true });
}

function parseSkill(skillDirName) {
    const skillMdPath = path.join(SKILLS_DIR, skillDirName, 'SKILL.md');
    if (!fs.existsSync(skillMdPath)) {
        return null;
    }

    const content = fs.readFileSync(skillMdPath, 'utf8');

    // Extract frontmatter if it exists
    const frontmatterRegex = /^---\s*\n([\s\S]*?)\n---\s*\n/;
    const match = content.match(frontmatterRegex);

    let name = skillDirName;
    let description = '';

    if (match) {
        const frontmatter = match[1];

        // Attempt to extract name
        const nameMatch = frontmatter.match(/^name:\s*(.+)$/m);
        if (nameMatch) name = nameMatch[1].trim();

        // Attempt to extract description
        const descMatchString = frontmatter.match(/^description:\s*(['"]?)([\s\S]*?)\1(?=\n[a-z_]+:|\s*$)/mi);
        if (descMatchString) {
            description = descMatchString[2].trim();
        }

        // Fallback description search if not in frontmatter
        if (!description) {
            const rest = content.slice(match[0].length).trim();
            const firstLine = rest.split('\n').map(l => l.trim()).filter(l => l.length > 0 && !l.startsWith('#'))[0];
            if (firstLine) description = firstLine;
        }
    } else {
        // No frontmatter, extract first text
        const lines = content.split('\n');
        for (let l of lines) {
            l = l.trim();
            if (l && !l.startsWith('#')) {
                description = l;
                break;
            }
        }
    }

    // Also look for name in `# title` if no frontmatter name
    if (name === skillDirName) {
        const h1Match = content.match(/^#\s+(.+)$/m);
        if (h1Match) {
            name = h1Match[1].trim();
        }
    }

    return {
        id: skillDirName,
        name: name,
        description: description.substring(0, 1000), // truncate long descriptions to 1000 chars
        path: `skills/${skillDirName}/SKILL.md`,
        url: `file://${path.resolve(skillMdPath)}`
    };
}

function generateSkillMap() {
    console.log(`Reading skills from ${SKILLS_DIR}`);
    if (!fs.existsSync(SKILLS_DIR)) {
        console.error('Skills directory not found.');
        process.exit(1);
    }

    const entries = fs.readdirSync(SKILLS_DIR, { withFileTypes: true });
    const skills = [];

    for (const entry of entries) {
        if (entry.isDirectory()) {
            const skill = parseSkill(entry.name);
            if (skill) {
                skills.push(skill);
            }
        }
    }

    console.log(`Found ${skills.length} skills. Writing to outputs...`);

    const resultJson = JSON.stringify(skills, null, 2);

    ensureDirectoryExistence(OUTPUT_FILE_1);
    fs.writeFileSync(OUTPUT_FILE_1, resultJson);
    console.log(`Saved: ${OUTPUT_FILE_1}`);

    ensureDirectoryExistence(OUTPUT_FILE_2);
    fs.writeFileSync(OUTPUT_FILE_2, resultJson);
    console.log(`Saved: ${OUTPUT_FILE_2}`);

    console.log('Success! Skill knowledge base mapping generated.');
}

generateSkillMap();

File: scripts/parse-skills.py
#!/usr/bin/env python3
import os
import json
import yaml
import re
from pathlib import Path

SKILLS_DIR = "skills"
MAP_FILE_1 = "docs/architecture/codemaps/003-arch-skills-map.json"
MAP_FILE_2 = ".agents/knowledge/kb_skills_mapping.json"

def parse_skill(skill_dir):
    skill_path = Path(skill_dir)
    skill_md = skill_path / "SKILL.md"
    
    if not skill_md.exists():
        return None
    
    with open(skill_md, "r", encoding="utf-8") as f:
        content = f.read()

    metadata = {}
    
    # Extract YAML frontmatter
    match = re.match(r"^---\n(.*?)\n---", content, re.DOTALL)
    if match:
        try:
            frontmatter = yaml.safe_load(match.group(1))
            if isinstance(frontmatter, dict):
                metadata.update(frontmatter)
        except Exception as e:
            print(f"Warning: Failed to parse frontmatter in {skill_md}: {e}")
            pass
            
        content_after_frontmatter = content[match.end():].strip()
    else:
        content_after_frontmatter = content.strip()

    name = metadata.get("name")
    if not name:
        # Try to find a heading 1
        h1_match = re.search(r"^#\s+(.+)$", content_after_frontmatter, re.MULTILINE)
        if h1_match:
            name = h1_match.group(1).strip()
        else:
            name = skill_path.name
            
    description = metadata.get("description", "").strip()
    if not description:
        # Try to find the first paragraph after a heading
        paragraphs = re.split(r'\n\s*\n', content_after_frontmatter)
        for p in paragraphs:
            p = p.strip()
            if p and not p.startswith('#') and not p.startswith('---') and not p.startswith('```') and not p.startswith('|'):
                description = p
                # Remove newlines within the paragraph
                description = ' '.join(description.split())
                break

    return {
        "id": skill_path.name,
        "name": name,
        "description": description,
        "directory": str(skill_path)
    }

def main():
    skills = []
    
    for item in os.listdir(SKILLS_DIR):
        skill_dir = os.path.join(SKILLS_DIR, item)
        if os.path.isdir(skill_dir):
            skill_data = parse_skill(skill_dir)
            if skill_data:
                skills.append(skill_data)
                
    # Sort alphabetically by ID
    skills.sort(key=lambda x: x["id"])
    
    os.makedirs(os.path.dirname(MAP_FILE_1), exist_ok=True)
    os.makedirs(os.path.dirname(MAP_FILE_2), exist_ok=True)

    with open(MAP_FILE_1, "w", encoding="utf-8") as f:
        json.dump(skills, f, indent=2)
        
    with open(MAP_FILE_2, "w", encoding="utf-8") as f:
        json.dump(skills, f, indent=2)
        
    print(f"Successfully mapped {len(skills)} skills.")
    print(f"Saved to {MAP_FILE_1} and {MAP_FILE_2}")

if __name__ == "__main__":
    main()
